test/wait-task/pkg/reconciler/reconciler_test.go

```diff
@@ -18,149 +18,247 @@ package reconciler
+		timeout: "1s",
+		params: `
+  params:
+  - name: duration
+    value: 2s
+`,
+		wantRunConditionType:   apis.ConditionSucceeded,
+		wantRunConditionStatus: corev1.ConditionFalse,
+		wantRunConditionReason: "ParentPipelineRunTimedOut",
+		isParentPRCancelled:    true,
 	}} {
-		t.Run(c.desc, func(t *testing.T) {
+		t.Run(tc.name, func(t *testing.T) {
 			ctx := context.Background()
 			rec := &Reconciler{}
 
+			runName := helpers.ObjectNameForTest(t)
+			runYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+    name: %s
+`, runName, tc.timeout, apiVersion, kind, tc.refName)
+			if tc.params != "" {
+				runYAML = runYAML + tc.params
+			}
+			r := mustParseRun(t, runYAML)
+			if tc.isParentPRCancelled {
+				r.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				r.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
 			// Start reconciling the Run.
 			// This will not return until the second Reconcile is done.
-			if err := rec.ReconcileKind(ctx, c.r); err != nil {
-				t.Fatalf("ReconcileKind() = %v", err)
+			err := rec.ReconcileKind(ctx, r)
+
+			t.Logf("%#v", r)
+
+			for err != nil {
+				if strings.Contains(err.Error(), "requeue") {
+					// simulate EnqueueAfter
+					var dur time.Duration
+					var dr error
+					for _, p := range r.Spec.Params {
+						if p.Name == "duration" {
+							dur, dr = time.ParseDuration(p.Value.StringVal)
+							if dr != nil {
+								t.Error("failed to parse duration")
+							}
+							break
+						}
+					}
+					to := r.GetTimeout()
+					if r.GetTimeout() == 0 {
+						to = config.DefaultTimeoutMinutes
+					}
+					sleep := int(math.Min(to.Seconds(), dur.Seconds()))
+					time.Sleep(time.Duration(sleep) * time.Second)
+					err = rec.ReconcileKind(ctx, r)
+				} else {
+					t.Fatalf("ReconcileKind() = %v", err)
+				}
 			}
 
-			if !c.r.IsDone() {
-				t.Fatal("Run was not done")
-			} else if c.r.IsSuccessful() {
-				t.Fatal("Run was successful")
+			// Compose expected Run
+			wantRunYAML := fmt.Sprintf(`
+metadata:
+  name: %s
+spec:
+  timeout: %s
+  ref:
+    apiVersion: %s
+    kind: %s
+    name: %s
+`, runName, tc.timeout, apiVersion, kind, tc.refName)
+			if tc.params != "" {
+				wantRunYAML = wantRunYAML + tc.params
+			}
+			wantRunYAML = wantRunYAML + fmt.Sprintf(`
+status:
+  conditions:
+  - reason: %s
+    status: %q
+    type: %s
+  observedGeneration: 0
+`, tc.wantRunConditionReason, tc.wantRunConditionStatus, tc.wantRunConditionType)
+			wantRun := mustParseRun(t, wantRunYAML)
+			if tc.isParentPRCancelled {
+				wantRun.Spec.Status = v1alpha1.RunSpecStatusCancelled
+				wantRun.Spec.StatusMessage = v1alpha1.RunCancelledByPipelineMsg
+			}
+
+			if d := cmp.Diff(wantRun, r,
+				filterTypeMeta,
+				filterObjectMeta,
+				filterCondition,
+				filterRunStatus,
+			); d != "" {
+				t.Errorf("-got +want: %v", d)
 			}
 		})
 	}
 }
+
+func mustParseRun(t *testing.T, yaml string) *v1alpha1.Run {
```

Comment:
this function exists [here](https://github.com/tektoncd/pipeline/blob/cd8f4f07d033c51d059aee65d20ccb95b709692b/test/parse/yaml.go#L38)

Reply:
yeah I copied it from the file. This reconciler_test.go belongs to wait-task module while the yaml helpers is outside of this wait-task/ and belongs to pipeline. Would it be better to make the two modules decoupled as much as possible?

Reply:
I think in this instance it's better to use the helper than have duplicate code. I don't see a benefit to having the wait task reconciler use separate code from the regular reconciler, as this wait task reconciler is used only in tekton pipelines for integration testing (can't be installed separately).

Reply:
makes sense, updated